File: apps/backend/nomad_vip/rank_contract.py
from __future__ import annotations

from datetime import date, timedelta
from decimal import Decimal
from typing import Any, Iterable, Mapping

from nomad_vip.daily_ranking import COMPONENT_ORDER, validate_thresholds
from nomad_vip.entertainer_ranks import (
	ACTIVE_ENTERTAINER_RANKS,
	payout_percent_for_rank,
)


COMPONENT_LABELS = {
	"attendance": "Ирц",
	"customer_complaints": "Зочны санал, гомдол",
	"sales": "Борлуулалт",
	"entertaining_skill": "Үзвэр, бүжгийн ур чадвар",
	"cleanliness_beauty": "Цэвэр байдал, төрх",
	"shift_effort": "Өдрийн гараа",
	"personal_development": "Хувийн хөгжил",
	"entertainer_attitude": "Хандлага",
}

SOURCE_LABELS = {
	"daily_absence": "Ирцийн бүртгэл",
	"absence_zero": "Тасалсан өдрийн 0 оноо",
	"absence_not_scored": "Тасалсан өдөр тооцохгүй",
	"normalized_event": "Баталгаажсан үнэлгээ",
	"attendance_normalization_required": "Ирцийн үнэлгээ хүлээгдэж байна",
	"seven_item_stage_round_checklist": "Өдрийн гарааны бүртгэл",
	"normalization_required": "Үнэлгээ хүлээгдэж байна",
	"latest_approved_assessment": "Сүүлийн баталгаажсан үнэлгээ",
	"approved_assessment_required": "Үнэлгээ хүлээгдэж байна",
	"substantiated_daily_incident": "Тайлбартай үнэлгээ",
	"policy_default_no_substantiated_incident": "Зөрчил бүртгэгдээгүй",
	"demo_batch": "Туршилтын өгөгдөл",
}

RANK_LABELS_MN = {
	"Rank 1": "1-р зэрэг",
	"Rank 2": "2-р зэрэг",
	"Rank 3": "3-р зэрэг",
}


def rank_label(rank: str | None) -> str:
	return RANK_LABELS_MN.get(str(rank or ""), str(rank or "—"))


def _date(value: Any) -> date | None:
	if isinstance(value, date):
		return value
	try:
		return date.fromisoformat(str(value)[:10])
	except (TypeError, ValueError):
		return None


def _number(value: Any) -> float | None:
	if value is None:
		return None
	return float(Decimal(str(value)))


def next_rank(rank: str | None) -> str | None:
	return {
		"Rank 3": "Rank 2",
		"Rank 2": "Rank 1",
	}.get(str(rank or ""))


def threshold_for_rank(rank: str | None, thresholds: Mapping[str, Any]) -> float | None:
	limits = validate_thresholds(thresholds)
	key = {
		"Rank 1": "rank_1",
		"Rank 2": "rank_2",
		"Rank 3": "rank_3",
	}.get(str(rank or ""))
	return float(limits[key]) if key else None


def payout_rules(thresholds: Mapping[str, Any]) -> list[dict[str, Any]]:
	limits = validate_thresholds(thresholds)
	return [
		{
			"rank": "Rank 3",
			"label": rank_label("Rank 3"),
			"minimum_score": 0.0,
			"maximum_score": float(limits["rank_2"]),
			"maximum_inclusive": False,
			"payout_percent": payout_percent_for_rank("Rank 3"),
		},
		{
			"rank": "Rank 2",
			"label": rank_label("Rank 2"),
			"minimum_score": float(limits["rank_2"]),
			"maximum_score": float(limits["rank_1"]),
			"maximum_inclusive": False,
			"payout_percent": payout_percent_for_rank("Rank 2"),
		},
		{
			"rank": "Rank 1",
			"label": rank_label("Rank 1"),
			"minimum_score": float(limits["rank_1"]),
			"maximum_score": 100.0,
			"maximum_inclusive": True,
			"payout_percent": payout_percent_for_rank("Rank 1"),
		},
	]


def component_contract(rows: Iterable[Mapping[str, Any]], thresholds: Mapping[str, Any]) -> list[dict[str, Any]]:
	by_key = {str(row.get("component")): row for row in rows}
	component_target = float(validate_thresholds(thresholds)["rank_3"])
	result = []
	for key in COMPONENT_ORDER:
		row = by_key.get(key, {})
		score = _number(row.get("score"))
		data_status = "verified" if row.get("status") == "verified" else ("not_applicable" if row.get("status") == "excluded" else "missing")
		target_status = "unknown" if score is None else ("met" if score >= component_target else "not_met")
		source = row.get("source") if isinstance(row.get("source"), Mapping) else {}
		result.append({
			"key": key,
			"label": COMPONENT_LABELS[key],
			"score": score,
			"weight": _number(row.get("weight")) or 0.0,
			"contribution": _number(row.get("contribution")),
			"data_status": data_status,
			"target_status": target_status,
			"source_label": SOURCE_LABELS.get(str(source.get("mode") or ""), "Баталгаажуулалт хүлээгдэж байна"),
		})
	return result


def build_rank_contract(
	*,
	snapshot: Mapping[str, Any] | None,
	effective_rank: str,
	effective_from: Any = None,
	thresholds: Mapping[str, Any],
	history: Iterable[Mapping[str, Any]] = (),
) -> dict[str, Any]:
	"""Build the API presentation contract without recalculating the daily score."""
	snapshot = snapshot or {}
	status = str(snapshot.get("status") or "Incomplete")
	complete = status == "Complete" and snapshot.get("weighted_score") is not None
	score = _number(snapshot.get("displayed_score") if snapshot.get("displayed_score") is not None else snapshot.get("weighted_score")) if complete else None
	calculated = str(snapshot.get("calculated_rank") or "") if complete else None
	scoring_date = _date(snapshot.get("scoring_date"))
	data_provenance = snapshot.get("input_provenance") or ("VERIFIED" if snapshot else "UNRESOLVED")

	promotion_rank = next_rank(effective_rank)
	promotion_threshold = threshold_for_rank(promotion_rank, thresholds)
	missing_score = None if score is None or promotion_threshold is None else max(0.0, round(promotion_threshold - score, 2))
	next_effective_from = (
		scoring_date + timedelta(days=1)
		if data_provenance != "DEMO"
		and scoring_date
		and calculated in ACTIVE_ENTERTAINER_RANKS
		and calculated != effective_rank
		else None
	)

	return {
		"scoring_date": scoring_date.isoformat() if scoring_date else None,
		"score": score,
		"daily_score": _number(snapshot.get("daily_score") if snapshot.get("daily_score") is not None else snapshot.get("weighted_score")) if complete else None,
		"counted_days": int(snapshot.get("counted_days") or 0),
		"score_basis": snapshot.get("score_basis") or "attendance_day_career_average",
		"score_status": "complete" if complete else "incomplete",
		"effective_rank": effective_rank,
		"effective_rank_label": rank_label(effective_rank),
		"effective_from": _date(effective_from).isoformat() if _date(effective_from) else None,
		"payout_percent": payout_percent_for_rank(effective_rank),
		"calculated_next_rank": calculated,
		"calculated_next_rank_label": rank_label(calculated) if calculated else None,
		"calculated_next_payout_percent": payout_percent_for_rank(calculated) if calculated in ACTIVE_ENTERTAINER_RANKS else None,
		"next_effective_from": next_effective_from.isoformat() if next_effective_from else None,
		"next_rank": promotion_rank,
		"next_rank_label": rank_label(promotion_rank) if promotion_rank else None,
		"next_rank_threshold": promotion_threshold,
		"missing_score": missing_score,
		"components": component_contract(snapshot.get("components") or (), thresholds),
		"missing_components": list(snapshot.get("missing_components") or ()),
		"history": list(history),
		"rules": payout_rules(thresholds),
		"data_provenance": data_provenance,
		"demo_batch": snapshot.get("demo_batch"),
	}

File: apps/entertainer/src/rankContract.ts
import type { DailyRankComponentKey, EntertainerDashboard, RankData } from './api'

type UnknownRecord = Record<string, unknown>

const COMPONENT_ORDER: DailyRankComponentKey[] = [
  'attendance',
  'customer_complaints',
  'sales',
  'entertaining_skill',
  'cleanliness_beauty',
  'shift_effort',
  'personal_development',
  'entertainer_attitude',
]

const COMPONENT_LABELS: Record<DailyRankComponentKey, string> = {
  attendance: 'Ирц',
  customer_complaints: 'Зочны санал, гомдол',
  sales: 'Борлуулалт',
  entertaining_skill: 'Үзвэр, бүжгийн ур чадвар',
  cleanliness_beauty: 'Цэвэр байдал, төрх',
  shift_effort: 'Өдрийн гараа',
  personal_development: 'Хувийн хөгжил',
  entertainer_attitude: 'Хандлага',
}

const DEFAULT_RULES: RankData['rules'] = [
  { rank: 'Rank 3', label: '3-р зэрэг', minimum_score: 0, maximum_score: 80, maximum_inclusive: false, payout_percent: 50 },
  { rank: 'Rank 2', label: '2-р зэрэг', minimum_score: 80, maximum_score: 90, maximum_inclusive: false, payout_percent: 60 },
  { rank: 'Rank 1', label: '1-р зэрэг', minimum_score: 90, maximum_score: 100, maximum_inclusive: true, payout_percent: 70 },
]

const asRecord = (value: unknown): UnknownRecord => (
  value && typeof value === 'object' && !Array.isArray(value) ? value as UnknownRecord : {}
)

const asRecords = (value: unknown): UnknownRecord[] => (
  Array.isArray(value) ? value.map(asRecord) : []
)

const text = (value: unknown): string | null => {
  if (typeof value !== 'string') return null
  const trimmed = value.trim()
  return trimmed || null
}

const numeric = (value: unknown): number | null => {
  if (value === null || value === undefined || value === '') return null
  const parsed = Number(value)
  return Number.isFinite(parsed) ? parsed : null
}

const rankName = (value: unknown): string => {
  const rank = text(value) || ''
  if (/^(rank\s*)?1$|1-\u0440|diamond|gold/i.test(rank)) return 'Rank 1'
  if (/^(rank\s*)?2$|2-\u0440|silver/i.test(rank)) return 'Rank 2'
  return 'Rank 3'
}

const rankLabel = (rank: string): string => (
  rank === 'Rank 1' ? '1-р зэрэг' : rank === 'Rank 2' ? '2-р зэрэг' : '3-р зэрэг'
)

const payoutPercent = (rank: string): number => (
  rank === 'Rank 1' ? 70 : rank === 'Rank 2' ? 60 : 50
)

const nextRank = (rank: string): string | null => (
  rank === 'Rank 3' ? 'Rank 2' : rank === 'Rank 2' ? 'Rank 1' : null
)

const nextThreshold = (rank: string | null): number | null => (
  rank === 'Rank 2' ? 80 : rank === 'Rank 1' ? 90 : null
)

const nextDate = (value: string | null): string | null => {
  if (!value) return null
  const date = new Date(`${value.slice(0, 10)}T00:00:00`)
  if (Number.isNaN(date.getTime())) return null
  date.setDate(date.getDate() + 1)
  return date.toISOString().slice(0, 10)
}

const componentKey = (value: unknown): DailyRankComponentKey | null => {
  const key = text(value)
  if (!key) return null
  if (COMPONENT_ORDER.includes(key as DailyRankComponentKey)) return key as DailyRankComponentKey
  if (key === 'behavior') return 'shift_effort'
  return null
}

/**
 * Accept both the current rank contract and the older production payload.
 * The adapter deliberately fills all eight rows so the rank view can never
 * crash merely because one backend field or list is absent.
 */
export function normalizeRankData(payload: unknown, fallback?: EntertainerDashboard): RankData {
  const root = asRecord(payload)
  const current = asRecord(root.current)
  const daily = Object.keys(asRecord(root.daily_rank)).length
    ? asRecord(root.daily_rank)
    : asRecord(fallback?.profile.daily_rank)
  const rawComponents = asRecords(root.components).length
    ? asRecords(root.components)
    : asRecords(daily.components).length
      ? asRecords(daily.components)
      : asRecords(root.evidence)
  const componentsByKey = new Map<DailyRankComponentKey, UnknownRecord>()
  rawComponents.forEach((row) => {
    const key = componentKey(row.key ?? row.component)
    if (key) componentsByKey.set(key, row)
  })

  const components: RankData['components'] = COMPONENT_ORDER.map((key) => {
    const row = componentsByKey.get(key) || {}
    const score = numeric(row.score ?? row.value)
    const status = text(row.data_status ?? row.status)
    const verified = status === 'verified'
    const notApplicable = status === 'not_applicable' || status === 'excluded'
    return {
      key,
      label: text(row.label) || COMPONENT_LABELS[key],
      score,
      weight: numeric(row.weight) || 0,
      contribution: numeric(row.contribution),
      data_status: verified ? 'verified' : notApplicable ? 'not_applicable' : 'missing',
      target_status: text(row.target_status) === 'met'
        ? 'met'
        : text(row.target_status) === 'not_met'
          ? 'not_met'
          : score == null
            ? 'unknown'
            : score >= 70 ? 'met' : 'not_met',
      source_label: text(row.source_label ?? row.detail) || (verified ? 'Баталгаажсан үнэлгээ' : 'Үнэлгээ хүлээгдэж байна'),
    }
  })

  const effectiveRank = rankName(root.effective_rank ?? current.current_rank ?? fallback?.profile.current_rank)
  const complete = text(root.score_status) === 'complete'
    || text(daily.status)?.toLowerCase() === 'complete'
  const score = complete
    ? numeric(root.score ?? daily.displayed_score ?? daily.weighted_score ?? current.current_points)
    : numeric(root.score)
  const calculated = complete && text(root.calculated_next_rank ?? daily.calculated_rank)
    ? rankName(root.calculated_next_rank ?? daily.calculated_rank)
    : null
  const promotionRank = text(root.next_rank)
    ? rankName(root.next_rank)
    : text(asRecord(root.next_target).rank)
      ? rankName(asRecord(root.next_target).rank)
      : nextRank(effectiveRank)
  const promotionThreshold = numeric(root.next_rank_threshold ?? asRecord(root.next_target).score_required)
    ?? nextThreshold(promotionRank)
  const scoringDate = text(root.scoring_date ?? daily.scoring_date)
  const missingComponents = Array.isArray(root.missing_components)
    ? root.missing_components.map(componentKey).filter((key): key is DailyRankComponentKey => Boolean(key))
    : Array.isArray(daily.missing_components)
      ? daily.missing_components.map(componentKey).filter((key): key is DailyRankComponentKey => Boolean(key))
      : components.filter((item) => item.data_status === 'missing').map((item) => item.key)

  const history: RankData['history'] = asRecords(root.history).map((row) => {
    const historicalRank = text(row.calculated_rank) ? rankName(row.calculated_rank) : null
    const followingRank = rankName(row.next_day_effective_rank ?? historicalRank ?? effectiveRank)
    return {
      scoring_date: text(row.scoring_date),
      score: numeric(row.score),
      score_status: text(row.score_status) === 'complete' ? 'complete' : 'incomplete',
      calculated_rank: historicalRank,
      calculated_rank_label: historicalRank ? rankLabel(historicalRank) : null,
      next_day_effective_rank: followingRank,
      next_day_effective_rank_label: rankLabel(followingRank),
      data_provenance: text(row.data_provenance) === 'DEMO'
        ? 'DEMO'
        : text(row.data_provenance) === 'VERIFIED'
          ? 'VERIFIED'
          : 'UNRESOLVED',
    }
  })

  const rules: RankData['rules'] = asRecords(root.rules).length
    ? asRecords(root.rules).map((row) => {
        const rank = rankName(row.rank)
        return {
          rank,
          label: text(row.label) || rankLabel(rank),
          minimum_score: numeric(row.minimum_score) || 0,
          maximum_score: numeric(row.maximum_score) ?? 100,
          maximum_inclusive: Boolean(row.maximum_inclusive),
          payout_percent: numeric(row.payout_percent) ?? payoutPercent(rank),
        }
      })
    : DEFAULT_RULES

  const calculatedChangesRank = Boolean(calculated && calculated !== effectiveRank)
  const missingScore = numeric(root.missing_score ?? asRecord(root.next_target).score_needed)
    ?? (score != null && promotionThreshold != null ? Math.max(0, promotionThreshold - score) : null)

  return {
    scoring_date: scoringDate,
    data_provenance: text(root.data_provenance) === 'DEMO'
      ? 'DEMO'
      : text(root.data_provenance) === 'VERIFIED'
        ? 'VERIFIED'
        : 'UNRESOLVED',
    demo_batch: text(root.demo_batch),
    score,
    daily_score: numeric(root.daily_score ?? daily.daily_score ?? daily.weighted_score),
    counted_days: numeric(root.counted_days ?? daily.counted_days) ?? 0,
    score_basis: 'attendance_day_career_average',
    score_status: score != null && complete ? 'complete' : 'incomplete',
    effective_rank: effectiveRank,
    effective_rank_label: text(root.effective_rank_label) || rankLabel(effectiveRank),
    effective_from: text(root.effective_from),
    payout_percent: numeric(root.payout_percent ?? current.payout_percent) ?? payoutPercent(effectiveRank),
    calculated_next_rank: calculated,
    calculated_next_rank_label: calculated ? rankLabel(calculated) : null,
    calculated_next_payout_percent: calculated
      ? (numeric(root.calculated_next_payout_percent)
        ?? rules.find(rule => rule.rank === calculated)?.payout_percent
        ?? payoutPercent(calculated))
      : null,
    next_effective_from: text(root.next_effective_from) || (calculatedChangesRank ? nextDate(scoringDate) : null),
    next_rank: promotionRank,
    next_rank_label: promotionRank ? rankLabel(promotionRank) : null,
    next_rank_threshold: promotionThreshold,
    missing_score: missingScore,
    missing_components: missingComponents,
    components,
    history,
    rules,
  }
}

export function rankDataFromDashboard(dashboard: EntertainerDashboard): RankData {
  return normalizeRankData({}, dashboard)
}
